lib/bye_bye_bye/utils.ex

```diff
@@ -0,0 +1,198 @@
+
+    %FeedEntity{
+      id: trip_id,
+      trip_update: %TripUpdate{
+        timestamp: now,
+        trip: %TripDescriptor{
+          trip_id: trip_id,
+          route_id: route_id,
+          schedule_relationship: "CANCELED"
+        },
+        stop_time_update: Enum.map(schedule, &schedule_to_stop_time_update/1)
+      }
+    }
+  end
+
+  defp schedule_to_stop_time_update(schedule) do
+    %StopTimeUpdate{
+      stop_id: schedule["relationships"]["stop"]["data"]["id"],
+      stop_sequence: schedule["attributes"]["stop_sequence"],
+      schedule_relationship: "SKIPPED"
+    }
+  end
+
+  @doc """
+  Gets schedules affected by a cancellation alert.
+  Returns a map of trip_id to schedule entries.
+
+  ## Parameters
+    * `alert` - The alert to process
+    * `now` - Current time as DateTime, used to determine service day
+  """
+  def get_affected_schedules(alert, now) do
+    period_params =
+      alert["attributes"]["active_period"]
+      |> Enum.map(&period_params(&1, now))
+      |> Enum.filter(& &1)
+
+    trips =
+      alert["attributes"]["informed_entity"]
+      |> Enum.filter(fn entity -> !entity["stop"] && entity["trip"] end)
+      |> Enum.map(& &1["trip"])
+      |> then(fn trips ->
+        case trips do
+          [] -> []
+          _ -> [%{trips: trips}]
+        end
+      end)
+
+    routes =
+      alert["attributes"]["informed_entity"]
+      |> Enum.filter(fn entity -> !entity["stop"] && !entity["trip"] && entity["route"] end)
+      |> Enum.map(& &1["route"])
+      |> then(fn routes ->
+        case routes do
+          [] -> []
+          _ -> [%{routes: routes}]
+        end
+      end)
+
+    params = trips ++ routes
+
+    params
+    |> Enum.flat_map(&Enum.map(period_params, fn params -> Map.merge(params, &1) end))
+    |> Enum.map(&MbtaClient.get_schedules/1)
+    |> Enum.map(fn {:ok, schedules} -> schedules end)
+    |> Enum.reduce(%{}, &Map.merge(&1, &2))
+  end
+
+  @doc """
+  Calculates the intersection of two time periods represented as DateTime tuples.
+
+  ## Parameters
+    * `{start1, end1}` - First time period as a tuple of DateTime objects
+    * `{start2, end2}` - Second time period as a tuple of DateTime objects
+
+  ## Returns
+    * `{intersection_start, intersection_end}` - The overlapping period if the periods intersect
+    * `nil` - If the periods do not intersect
+
+  ## Examples
+
+      iex> start1 = ~U[2024-01-01T10:00:00Z]
+      iex> end1 = ~U[2024-01-01T14:00:00Z]
+      iex> start2 = ~U[2024-01-01T12:00:00Z]
+      iex> end2 = ~U[2024-01-01T16:00:00Z]
+      iex> Utils.period_intersection({start1, end1}, {start2, end2})
+      {~U[2024-01-01T12:00:00Z], ~U[2024-01-01T14:00:00Z]}
+  """
+  def period_intersection({start1, end1}, {start2, end2}) do
+    latest_start = max(start1, start2)
+    earliest_end = min(end1, end2)
+
+    if latest_start <= earliest_end do
+      {latest_start, earliest_end}
+    else
+      nil
+    end
+  end
```

Comment:
Need to use the `DateTime` comparison functions rather than `max`, `min`, and `<=`, since those use structural comparison:
> Remember, comparisons in Elixir using [==/2](https://hexdocs.pm/elixir/Kernel.html#==/2), [>/2](https://hexdocs.pm/elixir/Kernel.html#%3E/2), [</2](https://hexdocs.pm/elixir/Kernel.html#%3C/2) and friends are structural and based on the DateTime struct fields. For proper comparison between datetimes, use the [compare/2](https://hexdocs.pm/elixir/DateTime.html#compare/2) function.

```suggestion
  def period_intersection({start1, end1}, {start2, end2}) do
    latest_start = Enum.max([start1, start2], DateTime)
    earliest_end = Enum.min([end1, end2], DateTime)

    if DateTime.compare(latest_start, earliest_end) in [:lt, :eq] do
      {latest_start, earliest_end}
    else
      nil
    end
  end
```

Reply:
Good call, interesting it appeared to be working as expected.

Reply:
Oh, but I bet it would go very wrong in the face of DateTimes with different timezones.